File: ProcessoConcrete/app/src/main/java/br/com/guilherme/concrete/view/PullRequestActivity.java
package br.com.guilherme.concrete.view;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import br.com.guilherme.concrete.R;
import br.com.guilherme.concrete.adapter.PullRequestAdapter;
import br.com.guilherme.concrete.model.PullRequest;
import br.com.guilherme.concrete.presenter.PullRequestPresenter;

public class PullRequestActivity extends AppCompatActivity implements PullRequestPresenter.View {
    private RecyclerView listPulls;
    private ProgressBar progressBar;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_repositorio);

        Toolbar toolbar = findViewById(R.id.toolbar);
        listPulls = findViewById(R.id.list);
        progressBar = findViewById(R.id.loading_content);

        String nomeRepositorio = getIntent().getStringExtra("nomeRepositorio") != null
                ? getIntent().getStringExtra("nomeRepositorio")
                : getResources().getString(R.string.title_padrao_reqeuests);
        toolbar.setTitle(nomeRepositorio);

        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                onBackPressed();
            }
        });

        PullRequestPresenter presenter = new PullRequestPresenter(this);
        presenter.getAllPulls();
    }

    public static Intent getIntent(Context context) {
        return new Intent(context, PullRequestActivity.class);
    }

    @Override
    public void setRecyclerView(List<PullRequest> pullRequests) {
        progressBar.setVisibility(View.GONE);
        listPulls.setVisibility(View.VISIBLE);

        PullRequestAdapter adapter = new PullRequestAdapter(pullRequests, PullRequestActivity.this);
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getApplicationContext());

        listPulls.setLayoutManager(linearLayoutManager);
        listPulls.setAdapter(adapter);
        listPulls.addItemDecoration(new DividerItemDecoration(getApplicationContext(), DividerItemDecoration.VERTICAL));
    }

    @Override
    public void onError(String errorException) {
        progressBar.setVisibility(View.GONE);
        Toast.makeText(getApplicationContext(), errorException, Toast.LENGTH_LONG).show();
    }
}

File: app/src/main/java/matheusuehara/github/features/pullrequests/PullRequestActivity.kt
package matheusuehara.github.features.pullrequests

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.MenuItem
import android.view.View.GONE
import android.view.View.VISIBLE
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.snackbar.Snackbar.LENGTH_INDEFINITE
import kotlinx.android.synthetic.main.activity_pull_request.*
import matheusuehara.github.R
import matheusuehara.github.data.model.PullRequest
import matheusuehara.github.data.model.ViewStateModel
import matheusuehara.github.util.Constants.INTENT_REPOSITORY_NAME
import matheusuehara.github.util.Constants.INTENT_REPOSITORY_OWNER_NAME
import matheusuehara.github.util.Constants.PULL_REQUEST_STATE_CLOSED
import matheusuehara.github.util.Constants.PULL_REQUEST_STATE_OPEN
import org.koin.androidx.viewmodel.ext.android.viewModel
import java.util.*

class PullRequestActivity : AppCompatActivity(), PullRequestClickListener {

    private var adapter: PullRequestAdapter = PullRequestAdapter(ArrayList(), this)
    private val pullRequestViewModel: PullRequestViewModel by viewModel()
    private lateinit var layoutManager: LinearLayoutManager

    private var repositoryName: String = String()
    private var repositoryOwnerName: String = String()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_pull_request)

        intent.getStringExtra(INTENT_REPOSITORY_NAME)?.let {
            repositoryName = it
        }

        intent.getStringExtra(INTENT_REPOSITORY_OWNER_NAME)?.let {
            repositoryOwnerName = it
        }

        supportActionBar?.title = repositoryName
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setDisplayShowHomeEnabled(true)

        initPullRequests()
        initObservable()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressed()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun initPullRequests() {
        layoutManager = LinearLayoutManager(this)
        rvPullRequest.layoutManager = layoutManager
        rvPullRequest.adapter = adapter
        rvPullRequest.addItemDecoration(DividerItemDecoration(this, layoutManager.orientation))

    }

    private fun initObservable() {
        getPullRequests()
        this.lifecycle.addObserver(pullRequestViewModel)
        pullRequestViewModel.getPullRequests().observe(this, Observer { stateModel ->
            when (stateModel.status) {
                ViewStateModel.Status.LOADING -> {
                    progress_bar.visibility = VISIBLE
                }
                ViewStateModel.Status.SUCCESS -> {
                    progress_bar.visibility = GONE
                    stateModel.model?.let { pullrequests ->
                        if (pullrequests.isEmpty()) showEmptyPullRequestMessage()
                        else {
                            adapter.addPullRequests(pullrequests)
                            val closedPr = pullrequests.count { it.state == PULL_REQUEST_STATE_CLOSED}
                            val openPr = pullrequests.count { it.state == PULL_REQUEST_STATE_OPEN }
                            status.text = getString(R.string.status, openPr, closedPr)
                        }
                    }
                }
                ViewStateModel.Status.ERROR -> {
                    progress_bar.visibility = GONE
                    showNetworkError()
                }
            }
        })
    }

    private fun getPullRequests() {
        pullRequestViewModel.loadPullRequests(repositoryOwnerName, repositoryName)
    }

    private fun showNetworkError() {
        Snackbar.make(
                frame_layout,
                R.string.connection_error,
                LENGTH_INDEFINITE)
                .setAction(R.string.try_again) {
                    getPullRequests()
                }.show()
    }

    private fun showEmptyPullRequestMessage() {
        Snackbar.make(frame_layout,
                R.string.empty_result,
                LENGTH_INDEFINITE).show()
    }

    override fun onClick(pullRequest: PullRequest) {
        val i = Intent(Intent.ACTION_VIEW)
        i.data = Uri.parse(pullRequest.html_url)
        startActivity(i)
    }

}
